net.bobbo.player-controller/player_scripts/crouch.gd

```diff
@@ -33,4 +40,37 @@ func player_physics_process(delta: float):
 	var scale_difference = new_scale - previous_scale
 	
 	player.velocity.y += scale_difference * 0.5
-	player.height = new_scale
+	
+	# Make it so that the crouch can only go as low as regularly crouching and only high enough such that the player doesn't get stuck if an object is detected above them
+	player.height = clamp(new_scale, _max_player_height / 2, _height_above)
+	
+	# Get the 3D physics space from a Node3D like the player_pivot
+	var space_state = player.model_pivot.get_world_3d().direct_space_state
+	
+	# Calculate origin and ending of raycast
+	var origin = player.position
+	var end = player.position + Vector3(0, _max_player_height / 2, 0)
```

Comment:
`Vector3(0, _max_player_height / 2, 0)` will do what we want it to in this case - but I suggest using the PlayerController's `up_direction` parameter as a basis instead of just assuming that (0, 1, 0) is up.

Something like
```
var end = player.position + ( player.up_direction * ( _max_player_height / 2 ) )
```

Reply:
We do this bc we want to support the player controller being able to be rotated. Walk on walls type stuff :D

Reply:
Gotcha, I didn't notice that the player had a variable up direction. I'll change the code to account for this